[BUG] classmethods don't compile using the limited api

### Describe the bug

Hello everyone,

When compiling a class with a classmethod using the limited API I get the following error:

```
clang -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk -DCYTHON_LIMITED_API=1 -DPy_LIMITED_API=0x030B0000 -I/private/var/folders/l4/gpyzwv455hzbmbv78x05555h0000gn/T/build-env-ok8opy1u/include -I/opt/homebrew/opt/python@3.11/Frameworks/Python.framework/Versions/3.11/include/python3.11 -c limited_classmethod.cpp -o build/temp.macosx-14-arm64-cpython-311/limited_classmethod.o
limited_classmethod.cpp:6876:5: error: no matching function for call to 'Py_TYPE'
    __Pyx_PyHeapTypeObject_GC_Del(m);
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
limited_classmethod.cpp:971:26: note: expanded from macro '__Pyx_PyHeapTypeObject_GC_Del'
    PyTypeObject *type = Py_TYPE(obj);\
                         ^~~~~~~
/opt/homebrew/opt/python@3.11/Frameworks/Python.framework/Versions/3.11/include/python3.11/object.h:132:29: note: candidate function not viable: no known conversion from '__pyx_CyFunctionObject *' to 'PyObject *' (aka '_object *') for 1st argument
static inline PyTypeObject* Py_TYPE(PyObject *ob) {
                            ^
limited_classmethod.cpp:7396:9: error: use of undeclared identifier 'PyMethodDescrObject'; did you mean 'PyMethodDescr_Type'?
        PyMethodDescrObject *descr = (PyMethodDescrObject *)method;
        ^~~~~~~~~~~~~~~~~~~
        PyMethodDescr_Type
/opt/homebrew/opt/python@3.11/Frameworks/Python.framework/Versions/3.11/include/python3.11/descrobject.h:22:26: note: 'PyMethodDescr_Type' declared here
PyAPI_DATA(PyTypeObject) PyMethodDescr_Type;
                         ^
limited_classmethod.cpp:7396:39: error: use of undeclared identifier 'PyMethodDescrObject'; did you mean 'PyMethodDescr_Type'?
        PyMethodDescrObject *descr = (PyMethodDescrObject *)method;
                                      ^~~~~~~~~~~~~~~~~~~
                                      PyMethodDescr_Type
/opt/homebrew/opt/python@3.11/Frameworks/Python.framework/Versions/3.11/include/python3.11/descrobject.h:22:26: note: 'PyMethodDescr_Type' declared here
PyAPI_DATA(PyTypeObject) PyMethodDescr_Type;
                         ^
limited_classmethod.cpp:7396:60: error: expected expression
        PyMethodDescrObject *descr = (PyMethodDescrObject *)method;
                                                           ^
limited_classmethod.cpp:7396:30: error: use of undeclared identifier 'descr'
        PyMethodDescrObject *descr = (PyMethodDescrObject *)method;
                             ^
limited_classmethod.cpp:7400:32: error: use of undeclared identifier 'descr'
        PyTypeObject *d_type = descr->d_common.d_type;
                               ^
limited_classmethod.cpp:7402:47: error: use of undeclared identifier 'descr'
        return PyDescr_NewClassMethod(d_type, descr->d_method);
                                              ^
limited_classmethod.cpp:7405:14: error: use of undeclared identifier 'PyMethod_Check'
    else if (PyMethod_Check(method)) {
             ^
limited_classmethod.cpp:7406:34: error: use of undeclared identifier 'PyMethod_GET_FUNCTION'
        return PyClassMethod_New(PyMethod_GET_FUNCTION(method));
                                 ^
limited_classmethod.cpp:7409:16: error: use of undeclared identifier 'PyClassMethod_New'
        return PyClassMethod_New(method);
               ^
10 errors generated.

```

### Code to reproduce the behaviour:

```cython
# limited_classmethod.pyx

cdef class A:
    @classmethod
    def method(cls):
        pass
```


### Expected behaviour

It should compile.

### OS

_No response_

### Python version

_No response_

### Cython version

_No response_

### Additional context

classmethod implementation: https://github.com/cython/cython/blob/0f2a0d38b60e10e401aa1435801b2b8975083c7b/Cython/Utility/CythonFunction.c#L1754-L1804

--- a/Cython/Utility/CythonFunction.c
+++ b/Cython/Utility/CythonFunction.c
@@ -1753,7 +1753,9 @@ static int __pyx_FusedFunction_init(PyObject *module) {
 
 //////////////////// ClassMethod.proto ////////////////////
 
+#if !CYTHON_COMPILING_IN_LIMITED_API
 #include "descrobject.h"
+#endif
 CYTHON_UNUSED static PyObject* __Pyx_Method_ClassMethod(PyObject *method); /*proto*/
 
 //////////////////// ClassMethod ////////////////////
@@ -1784,6 +1786,7 @@ static PyObject* __Pyx_Method_ClassMethod(PyObject *method) {
     if (__Pyx_TypeCheck(method, methoddescr_type))
 #endif
     {
+#if !CYTHON_COMPILING_IN_LIMITED_API
         // cdef classes
         PyMethodDescrObject *descr = (PyMethodDescrObject *)method;
         #if PY_VERSION_HEX < 0x03020000
@@ -1792,8 +1795,17 @@ static PyObject* __Pyx_Method_ClassMethod(PyObject *method) {
         PyTypeObject *d_type = descr->d_common.d_type;
         #endif
         return PyDescr_NewClassMethod(d_type, descr->d_method);
+#else
+        return PyErr_Format(
+            PyExc_SystemError,
+            "Cython cannot yet handle classmethod on a MethodDescriptorType (%S) in limited API mode. "
+            "This is most likely a classmethod in a cdef class method with binding=False. "
+            "Try setting 'binding' to True.",
+            method);
+#endif
     }
 #endif
+#if !CYTHON_COMPILING_IN_LIMITED_API
     else if (PyMethod_Check(method)) {
         // python classes
         return PyClassMethod_New(PyMethod_GET_FUNCTION(method));
@@ -1801,4 +1813,36 @@ static PyObject* __Pyx_Method_ClassMethod(PyObject *method) {
     else {
         return PyClassMethod_New(method);
     }
+#else
+    {
+        PyObject *types_module, *method_type=NULL, *func=NULL;
+        PyObject *builtins, *classmethod, *result=NULL;
+        types_module = PyImport_ImportModule("types");
+        if (!types_module) {
+            return NULL;
+        }
+        method_type = PyObject_GetAttrString(types_module, "MethodType");
+        if (!method_type) goto bad;
+        if (__Pyx_TypeCheck(method, method_type)) {
+            func = PyObject_GetAttrString(method, "__func__");
+            if (!func) goto bad;
+        } else {
+            func = method;
+            Py_INCREF(func);
+        }
+        builtins = PyEval_GetBuiltins(); // borrowed
+        if (!builtins) goto bad;
+        classmethod = PyDict_GetItemString(builtins, "classmethod");
+        // classmethod is borrowed
+        if (!classmethod) goto bad;
+        result = PyObject_CallFunctionObjArgs(classmethod, func, NULL);
+
+        bad:
+        Py_XDECREF(func);
+        Py_XDECREF(method_type);
+        Py_DECREF(types_module);
+        return result;
+    }
+#endif
+
 }

--- a/tests/run/isolated_limited_api_tests.srctree
+++ b/tests/run/isolated_limited_api_tests.srctree
@@ -38,6 +38,9 @@ limited.C()
 limited.D()
 limited.E()
 
+assert limited.C.cm() == limited.C().cm() == limited.C
+assert limited.D.cm() == limited.D().cm() == limited.D
+
 assert limited.decode(b'a', bytearray(b'b')) == "ab"
 
 assert limited.cast_float(1) == 1.0
@@ -82,10 +85,14 @@ def cast_float(object o):
     return float(o)
 
 class C:
-    pass
+    @classmethod
+    def cm(cls):
+        return cls
 
 cdef class D:
-    pass
+    @classmethod
+    def cm(cls):
+        return cls
 
 cdef class E(D):
     pass

